src/main/scala/essential_scala/Ex3Classes.scala

```diff
@@ -17,9 +17,117 @@ object Ex3Classes extends App {
     def willServe(cat: Cat1) = cat.food == "Chips"
   }
 
+  val eastwood = new Director("Clint", "Eastwood", 1930)
+  val mcTiernan = new Director("John", "McTiernan", 1951)
+  val nolan = new Director("Christopher", "Nolan", 1970)
+  val someBody = new Director("Just", "SomeBody", 1990)
+  val memento = new Film("Memento", 2000, 8.5, nolan)
+  val darkKnight = new Film("DarkKnight", 2008, 9.0, nolan)
+  val inception = new Film("Inception", 2010, 8.8, nolan)
+  val highPlainsDrifter = new Film("HighPlainsDrifter", 1973, 7.7, eastwood)
+  val outlawJoseyWales = new Film("TheOutlawJoseyWales", 1976, 7.9, eastwood)
+  val unforgiven = new Film("Unforgiven", 1992, 8.3, eastwood)
+  val granTorino = new Film("GranTorino", 2008, 8.2, eastwood)
+  val invictus = new Film("Invictus", 2009, 7.4, eastwood)
+  val predator = new Film("Predator", 1987, 7.9, mcTiernan)
+  val dieHard = new Film("DieHard", 1988, 8.3, mcTiernan)
+  val huntForRedOctober = new Film("TheHuntforRedOctober", 1990, 7.6, mcTiernan)
+  val thomasCrownAffair = new Film("TheThomasCrownAffair", 1999, 6.8, mcTiernan)
+  println(eastwood.yearOfBirth)
+  println(dieHard.director.name)
+  println(invictus.isDirectedBy(nolan))
+
+  println(highPlainsDrifter.copy(name = "L'homme des hautes plaines"))
+  println(
+    thomasCrownAffair.copy(yearOfRealease = 1968,
+                           director = new Director("Norman", "Jewison", 1926)))
+  println(inception.copy().copy().copy())
+
+  // 3.1.6.4 A Simple Counter
+  // 3.1.6.6 Additional Counting
+  class Adder(amount: Int) {
+    def add(in: Int) = in + amount
+  }
+  class Counter(val count: Int) {
+    def inc = new Counter(count + 1)
+    def dec = new Counter(count - 1)
+    def adjust(adder: Adder): Counter = {
+      new Counter(adder.add(count))
+    }
+  }
+  println(new Counter(10).inc.dec.inc.inc.count)
+  println(new Counter(10).adjust(new Adder(4)).count)
+  // 3.1.6.5 Counting Faster
+  class Counter1(val count: Int) {
+    def inc: Counter1 = inc()
+    def dec: Counter1 = dec()
+    def inc(v: Int = 1): Counter1 = new Counter1(count + v)
+    def dec(v: Int = 1): Counter1 = new Counter1(count - v)
+  }
+  println(new Counter1(10).inc.inc(10).count)
+
+  // 3.2.3.1 When is a Function is not a Function?
+  // other developers need to know about our specific class to use it
+
   //  3.4.5.1 Case Cats
   case class Cat2(
     colour: String,
     food: String
   )
 }
+
+// 3.3.2.1 Friendly Person Factory
+class Person(firstName: String, lastName: String)
+object Person {
+  def apply(name: String): Person = {
+    val parts = name.split(" ")
+    new Person(parts(0), parts(1))
```

Comment:
What if I call it as `Person("Chingachgook")`, or `Person("Chingachgook The Great Snake")`?

Reply:
`Person("Chingachgook")` lead to
`Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1`, but `Person("Chingachgook The Great Snake")` lead to lost data in Person "Chingachgook The".

We can handle that tricky cases to.